From a0c292275dbf6a2efc9b6a08e61391aa0973283b Mon Sep 17 00:00:00 2001
From: Jaroslav Kysela <jkysela@redhat.com>
Date: Wed, 3 Sep 2014 07:20:20 -0400
Subject: [alsa] hda/ca0132: setup/cleanup streams

Message-id: <140972823573-128-patchset-jkysela@redhat.com>
Patchwork-id: 90828
O-Subject: [RHEL 7.1 PATCH 127/442] ALSA: hda/ca0132 - setup/cleanup streams
Bugzilla: 1112200
RH-Acked-by: John Linville <linville@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Ivan Vecera <ivecera@redhat.com>

When a HDMI stream is opened with the same stream tag
as a following opened stream to ca0132, audio will be
heard from two ports simultaneously.
Fix this issue by change to use snd_hda_codec_setup_stream
and snd_hda_codec_cleanup_stream instead, so that an
inactive stream can be marked as 'dirty' when found
with a conflict stream tag, and then get purified.

Signed-off-by: Hsin-Yu Chao <hychao@chromium.org>
Reviewed-by: Chih-Chung Chang <chihchung@chromium.org>
Cc: <stable@vger.kernel.org>
Signed-off-by: Takashi Iwai <tiwai@suse.de>

(picked from kernel: 28fba95087a7f3d107a3a6728aef7dbfaf3fd782)
rhbz#1112200
Signed-off-by: Jaroslav Kysela <jkysela@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/sound/pci/hda/patch_ca0132.c b/sound/pci/hda/patch_ca0132.c
index 4c5132c..b9f2cbb 100644
--- a/sound/pci/hda/patch_ca0132.c
+++ b/sound/pci/hda/patch_ca0132.c
@@ -2661,60 +2661,6 @@ static bool dspload_wait_loaded(struct hda_codec *codec)
 }
 
 /*
- * PCM stuffs
- */
-static void ca0132_setup_stream(struct hda_codec *codec, hda_nid_t nid,
-     u32 stream_tag,
-     int channel_id, int format)
-{
- unsigned int oldval, newval;
-
- if (!nid)
-  return;
-
- snd_printdd(
-     "ca0132_setup_stream: NID=0x%x, stream=0x%x, "
-     "channel=%d, format=0x%x\n",
-     nid, stream_tag, channel_id, format);
-
- /* update the format-id if changed */
- oldval = snd_hda_codec_read(codec, nid, 0,
-        AC_VERB_GET_STREAM_FORMAT,
-        0);
- if (oldval != format) {
-  msleep(20);
-  snd_hda_codec_write(codec, nid, 0,
-        AC_VERB_SET_STREAM_FORMAT,
-        format);
- }
-
- oldval = snd_hda_codec_read(codec, nid, 0, AC_VERB_GET_CONV, 0);
- newval = (stream_tag << 4) | channel_id;
- if (oldval != newval) {
-  snd_hda_codec_write(codec, nid, 0,
-        AC_VERB_SET_CHANNEL_STREAMID,
-        newval);
- }
-}
-
-static void ca0132_cleanup_stream(struct hda_codec *codec, hda_nid_t nid)
-{
- unsigned int val;
-
- if (!nid)
-  return;
-
- snd_printdd(KERN_INFO "ca0132_cleanup_stream: NID=0x%x\n", nid);
-
- val = snd_hda_codec_read(codec, nid, 0, AC_VERB_GET_CONV, 0);
- if (!val)
-  return;
-
- snd_hda_codec_write(codec, nid, 0, AC_VERB_SET_STREAM_FORMAT, 0);
- snd_hda_codec_write(codec, nid, 0, AC_VERB_SET_CHANNEL_STREAMID, 0);
-}
-
-/*
  * PCM callbacks
  */
 static int ca0132_playback_pcm_prepare(struct hda_pcm_stream *hinfo,
@@ -2725,7 +2671,7 @@ static int ca0132_playback_pcm_prepare(struct hda_pcm_stream *hinfo,
 {
  struct ca0132_spec *spec = codec->spec;
 
- ca0132_setup_stream(codec, spec->dacs[0], stream_tag, 0, format);
+ snd_hda_codec_setup_stream(codec, spec->dacs[0], stream_tag, 0, format);
 
  return 0;
 }
@@ -2744,7 +2690,7 @@ static int ca0132_playback_pcm_cleanup(struct hda_pcm_stream *hinfo,
  if (spec->effects_switch[PLAY_ENHANCEMENT - EFFECT_START_NID])
   msleep(50);
 
- ca0132_cleanup_stream(codec, spec->dacs[0]);
+ snd_hda_codec_cleanup_stream(codec, spec->dacs[0]);
 
  return 0;
 }
@@ -2823,8 +2769,8 @@ static int ca0132_capture_pcm_prepare(struct hda_pcm_stream *hinfo,
 {
  struct ca0132_spec *spec = codec->spec;
 
- ca0132_setup_stream(codec, spec->adcs[substream->number],
-       stream_tag, 0, format);
+ snd_hda_codec_setup_stream(codec, spec->adcs[substream->number],
+       stream_tag, 0, format);
 
  return 0;
 }
@@ -2838,7 +2784,7 @@ static int ca0132_capture_pcm_cleanup(struct hda_pcm_stream *hinfo,
  if (spec->dsp_state == DSP_DOWNLOADING)
   return 0;
 
- ca0132_cleanup_stream(codec, hinfo->nid);
+ snd_hda_codec_cleanup_stream(codec, hinfo->nid);
  return 0;
 }
 
@@ -4741,6 +4687,8 @@ static int patch_ca0132(struct hda_codec *codec)
   return err;
 
  codec->patch_ops = ca0132_patch_ops;
+ codec->pcm_format_first = 1;
+ codec->no_sticky_stream = 1;
 
  return 0;
 }
-- 
1.7.1